### PR TITLE
paramset add table helper

### PR DESCRIPTION
**Pull request by tehn**
fixes https://github.com/monome/norns/issues/569

### CODE REVIEW

File: lua/paramset.lua [hunk 1]
```diff
@@ -40,11 +40,43 @@ end
 
 --- add generic parameter
```
Comment (pq) [inline lua/paramset.lua]: perhaps some docs here?

File: lua/paramset.lua [hunk 2]
```diff
@@ -40,11 +40,43 @@ end
...
 function ParamSet:add(args)
-  local param = args.param -- param is mandatory
+  local param = args.param
+  if param == nil then
+    if args.number then
+      local id = args.number
+      local name = args.name or id
+      param = number.new(id, name, args.min, args.max, args.default)
```
Comment (pq) [inline lua/paramset.lua]: in general, i wonder if, in addition to docs, we don't want to some validation to ensure required args are provided.

if we had something like

```lua
local function checkRequired(args, names) {
  for _,name in names do
    if args.name == nil then error('argument "'..name..'" is required')
  end
} 
```

we could do something like:

```lua
  checkRequired(args, {'min', 'max'});
```

which should report something useful like `argument "max" is required` (only better) if someone did something like:

```lua
params:add{number = 'foo', min = 0}
```

(caveat: all untested; just food for thought!)

File: lua/paramset.lua [hunk 3]
```diff
@@ -40,11 +40,43 @@ end
...
+    elseif args.taper then
+      local id = args.taper
+      local name = args.name or id
+      param = taper.add(id, name, args.min, args.max, args.default, args.k, args.units)
+    elseif args.trigger then
+      local id = args.trigger
+      local name = args.name or id
+      param = trigger.add(id, name)
+    else
+      print("paramset.add() error")
```
Comment (pq) [inline lua/paramset.lua]: how about something like:

```lua
print('paramset.add() error: unspecified param type (expected   "option", "control", "file", ... ')
```

**Comment by artfwo** [comment]
i might have missed some earlier discussions, but why is `add` taking `number`, `trigger`, etc. arguments? why not `add{type='number', id='foo'}`?

**Comment by tehn** [comment]
> why not` add{type='number', id='foo'}`?

this is probably more straightforward and normal, though more verbose of course. i think the goal was to make a concise syntax. @pq @artfwo

File: lua/paramset.lua (hunk 2 above)
Comment (tehn) [reply lua/paramset.lua]: this sounds nice, certainly. but most fields really aren't required... they just need sensible defaults. have you scanned through to see if there are many required fields?

**Comment by pq** [comment]
> this is probably more straightforward and normal, though more verbose of course.

yeah.  i like it concise but maybe it's a bridge too far.

curious what other folks think.

/cc @markwheeler @antonhornquist @simonvanderveldt

File: lua/paramset.lua [hunk 4]
```diff
@@ -39,12 +39,46 @@ function ParamSet:add_separator()
...
+      return nil
+    elseif args.id == nil then
+      print("paramset.add() error: id required")
+      return nil
+    end
+
+    local id = args.id
+    local name = args.name or id
+
+    if args.number then
```
Comment (artfwo) [inline lua/paramset.lua]: hmm, but shouldn't we check args.type here instead?

Comment (tehn) [reply lua/paramset.lua]: sorry totally messed this up

**Pull request merged**